### PR TITLE
[Input] Final Fixes (MMJOY + evdev)

### PR DESCRIPTION
**Pull request by Megamouse**
#3796 
#3818
#3825
#3839 
#3924

### CODE REVIEW

**Comment by flash-fire** [comment]
If this is actually your last PR for this, I promise to shave my head.

**Comment by Psycho-A** [comment]
Will be plan to implement mouse axis as some of gamepad's analog sticks? It would be great in the end! :)

**Comment by Psycho-A** [comment]
Found first bugs with this PR.
Keyboard now is totally buggy. I binded Shift as R1 button, and WASD to Left stick, and now I can't control WASD keys when shift is pressed. Probably the same thing happen with some other buttons too. Also, low fps now borns wild input lags with button stucking, f.e. my character was forced to move left when left direction on stick were released during low fps.

**Comment by Megamouse** [comment]
try again @Psycho-A

**Comment by Psycho-A** [comment]
@Megamouse 
works well as before.

**Comment by toccata10** [comment]
Hi Megamouse,
A bit of testing with the evdev part:
*with this PR: ok for an xbox360 controller. Vibrations are cool too.

With my custom stick:
*on master: works fine except that the analog axis are reversed: it's ok on the config window, but when in a game: if I press up, it goes down. If I press right, it goes left.
Correct in gui window, wrong in games:
selection_043

Workaround: I press left to register right input in the config window.
Dpad is perfect.
Incorrect in gui window, correct in games:
selection_044

So, while not perfect, it's very easy to configure.

*on PR: I can't register LX-, neither LY-: it always registers LZ+
So it seems like LX- and LY- are not captured.
Knowing that with an evtest, this is my output (so if I don't press anything, the gui always captures a LZ+):
```
Event: time 1512561543.089711, -------------- SYN_REPORT ------------
Event: time 1512561543.097704, type 3 (EV_ABS), code 2 (ABS_Z), value 133
Event: time 1512561543.097704, -------------- SYN_REPORT ------------
Event: time 1512561543.105706, type 3 (EV_ABS), code 2 (ABS_Z), value 132
Event: time 1512561543.105706, -------------- SYN_REPORT ------------
Event: time 1512561543.113707, type 3 (EV_ABS), code 2 (ABS_Z), value 131
Event: time 1512561543.113707, -------------- SYN_REPORT ------------
Event: time 1512561543.121721, type 3 (EV_ABS), code 2 (ABS_Z), value 133
Event: time 1512561543.121721, -------------- SYN_REPORT ------------
Event: time 1512561543.129712, type 3 (EV_ABS), code 2 (ABS_Z), value 132
```
Sometimes, when I look at the analog deadzones diagram, there's a big offset for the center point. That's my picture after pressing the "restore defaults" (and moving the stick doesn't magically centers the rest point):
configure evdev_046


If I press on the filter noise button, it outputs this:
```
E HLE: Evdev Calibration: Added axis [ 5 = RZ+ ] to blacklist. Value = 127
E HLE: Evdev Calibration: Added axis [ 3 = RX+ ] to blacklist. Value = 127
E HLE: Evdev Calibration: Added axis [ 2 = LZ+ ] to blacklist. Value = 135
E HLE: Evdev Calibration: Added axis [ 1 = LY+ ] to blacklist. Value = 127
E HLE: Evdev Calibration: Added axis [ 0 = LX+ ] to blacklist. Value = 127
```
which is perfect for registering my buttons, but of course, doesn't allow me to input joystick signals anymore.
Suggestion: a multiselection drop-down box called "Axis Blacklisting" in which we can disable the problematic axis by ticking them instead of the "Filter noise" button.

Hope it's detailled enough to help you.

**Comment by Megamouse** [comment]
I'll add min and max size to the log.
You can do so as well if you know how to see what the range is.
If it is non negative then the problem is clear

**Comment by Megamouse** [comment]
Axis are checked by ranges from negative to positive space while triggers are checked by 0+ range.

So if your axis has 0+ range it will be handled as trigger.

I could do a button like "handle trigger axis" to whitelist those axis and handle them accordingly

**Comment by Megamouse** [comment]
I added min and max to the log message in order to make sure I'm doing the right thing next

**Comment by LuismaSP89** [comment]
@Megamouse  I opened an issue with imput gamepad, (https://github.com/RPCS3/rpcs3/issues/3924) i don´t know if this is related with your work in this PR. 

BTW, this issue happens in master build and with the latest build (artifact) of this PR in appveyor.

**Comment by Psycho-A** [comment]
When will this be merged? Keyboard is still broken on Master builds as I described it above...

**Comment by LuismaSP89** [comment]
@Psycho-A   Be patient, every PR needs a lot of testing and pass all checks. It's better to be sure that it doesn't break things.

**Comment by toccata10** [comment]
Checking for the extreme values, I added this in your code:
	```
case EV_ABS:
	{
		auto& dev = device.device;
		int min = libevdev_get_abs_minimum(dev, code);
		int max = libevdev_get_abs_maximum(dev, code);
std::cout <<"code EV_ABS "<< code<< "max " << max <<  std::endl;
std::cout << "code EV_ABS "<< code<<"min " << min <<  std::endl;

```
Here are the outputs:
xbox 360:
```
code EV_ABS 1max 32767
code EV_ABS 1min -32768
code EV_ABS 0max 32767
code EV_ABS 0min -32768
code EV_ABS 0max 32767
code EV_ABS 0min -32768
code EV_ABS 4max 32767
code EV_ABS 4min -32768
```
DragonRise:
```
code EV_ABS 2max 255
code EV_ABS 2min 0
code EV_ABS 2max 255
code EV_ABS 2min 0
code EV_ABS 2max 255

```

**Comment by toccata10** [comment]
I never tested with 2 joysticks after the nice GUI PR. On Linux, the 2nd stick is not usable: it's detected in the Configuration>Pads menu, but it's not detected in the game.
The log shows only "Opened joystick: 'Microsoft X-Box 360 pad' at /dev/input/event20 (fd 23)", never the 2nd one (on event19).
What works is:
*one player on evdev
*the other one on keyboard
May I suggest that in the Configuration>Pads menu, the evdev number appears (ex: Microsoft X-Box 360 pad (event19) and Microsoft X-Box 360 pad (event20) ) ? That could avoid some confusion with multiple contollers.
My log with 2 xbox sticks connected.
[RPCS3.log-2sticksPb.gz](https://github.com/RPCS3/rpcs3/files/1569435/RPCS3.log-2sticksPb.gz)
The last working version was:
rpcs3-v0.0.4-2017-11-23-7d3a528-8931_linux64 (just before the Gui configuration merge).

**Comment by Megamouse** [comment]
@toccata10 try this change please

image

**Comment by toccata10** [comment]
this renders rpcs3 very unstable (it segfaults).

File: rpcs3/Emu/Io/PadHandler.h [hunk 1]
```diff
@@ -319,6 +319,8 @@ class PadHandlerBase
 protected:
 	static const u32 MAX_GAMEPADS = 7;
 
+	std::array<bool, MAX_GAMEPADS> last_connection_status{{ false, false, false, false, false, false, false }};
```
Comment (elad335) [inline rpcs3/Emu/Io/PadHandler.h]: why not using a bitset instead of a bool array

**Comment by toccata10** [comment]
Commenting the return std:.. line and adding a continue works: I tested with 2 xbox sticks and it works fine.

File: rpcs3/Emu/Io/PadHandler.h (hunk 1 above)
Comment (rodjun) [reply rpcs3/Emu/Io/PadHandler.h]: For 7 boolean values? Seriously?

**Comment by Asinin3** [comment]
This Pull Request broke Catherine for me and another user with DS3 and DS4. After going to the title screen it won't let you load a game or open the configuration menu. You can move the cursor/selection up and down but pressing X won't do anything.

**Comment by spyropt** [comment]
This broke NPUB30034 Super Street Fighter 2 Turbo HD using mmjoystick x is not working in the game
everything else appears to work moving up and down, but start is mapped to select in this game(its ok in mmjoystick config) 
tried the latest build and its the same

**Comment by MDAshcreator** [comment]
Mmjoystick is not responding on rpcs3

**Pull request merged**